crates/spirv-std/src/arch.rs

```diff
@@ -27,3 +27,28 @@ pub unsafe fn vector_extract_dynamic<T: Scalar, V: Vector<T>>(vector: V, index:
 
     result
 }
+
+/// Make a copy of a vector, with a single, variably selected,
+/// component modified.
+///
+/// # Safety
+/// Behavior is undefined if `index`’s value is greater than or equal to the
+/// number of components in `vector`.
+#[spirv_std_macros::gpu_only]
+#[doc(alias = "OpVectorInsertDynamic")]
+#[inline]
+pub unsafe fn vector_insert_dynamic<T: Scalar, V: Vector<T>>(vector: V, index: usize, element: T) -> V {
+    let mut result = V::default();
+
+    asm! {
+        "%vector = OpLoad _ {vector}",
+        "%new_vector = OpVectorInsertDynamic _ %vector {element} {index}",
```

Comment:
@khyperia || @eddyb Would you happen to know why inference would fail here?

Reply:
Given this, I'm guessing the `FIXME`?
https://github.com/EmbarkStudios/rust-gpu/blob/2f9d4d8e0d9dce5d06154929025e2c1fc0182a71/crates/rustc_codegen_spirv/src/spirv_type_constraints.rs#L373-L376

It requires adding another type pattern variable and a type pattern like `Either` but which is "both must match at once" - a bit like Rust's `x @ Foo(_)` - I almost did it, IIRC for @msiglreith, but then got swamped further by #300 (the implementation of which also interacts with this but not in any difficult way).

Reply:
Should I try to add that, or wait till #414 is merged?

Reply:
> Should I try to add that

Feel free to! It should be easy for me to rebase on such a change.